src/spec.ts

```diff
@@ -319,9 +339,44 @@ export function alwaysNoOcclusion(spec: ExtendedUnitSpec): boolean {
   return vlEncoding.isAggregate(spec.encoding);
 }
 
-export function fieldDefs(spec: ExtendedUnitSpec): FieldDef[] {
+/* Accumulate non-duplicate fieldDefs in a dictionary */
```

Comment:
I'm sure there is a way to write these 3 methods in a more compact way as just one method.  
Right now the code seems unnecessarily verbose.  

Hint:

Add  `dict: Dict<FieldDef> = {}` as second parameter to `fieldDefs` and you can just reuse it.